Zend/Optimizer/dce.c

```diff
@@ -245,15 +246,9 @@ static inline bool may_have_side_effects(
 				if ((opline->extended_value & (ZEND_BIND_IMPLICIT|ZEND_BIND_EXPLICIT))) {
 					return 1;
 				}
-
-				if ((opline->extended_value & ZEND_BIND_REF) != 0) {
-					zval *value =
-						(zval*)((char*)op_array->static_variables->arData +
-							(opline->extended_value & ~ZEND_BIND_REF));
-					if (Z_TYPE_P(value) == IS_CONSTANT_AST) {
-						/* AST may contain undefined constants */
-						return 1;
-					}
+				/* Modifies local variables and static variables which are observable through reflection */
+				if (opline->extended_value & ZEND_BIND_REF) {
```

Comment:
Would the check for ZEND_JMP_STATIC_DEF suffice for checking for side effects? (See other comments on zend_compile.c)

And would opcodes such as `ZEND_ASSIGN` modifying the reference be triggered as causing `may_have_side_effects`.

(e.g. is it only necessary to check ZEND_BIND_STATIC for `IS_TYPE_UNINITIALIZED` here similar to the old code)

Reply:
@TysonAndre I might be missing the point of `may_have_side_effects`. `ZEND_BIND_STATIC` modifies local variables (even without `ZEND_BIND_REF`), doesn't this mean this always have side effects? What does "side effect" mean in the context of DCE?

Reply:
I took a closer look. Apparently declaring variables is not considered an intrinsic side effect. Instead, the usage of the variable is, which will cause the declaration to be marked as live. After this RFC, we no longer accept constant ASTs so that side effect goes away (which should've been in `may_throw` anyway). However, we do have a new case where `OP2` is assigned to the static variable which we need to account for.